nautobot_netbox_importer/generator/fields.py

```diff
@@ -65,22 +68,68 @@ def define_relation(field: SourceField) -> None:
     return define_relation
 
 
-def role(adapter: SourceAdapter, source_content_type: ContentTypeStr) -> SourceFieldDefinition:
+# To ensure all role models will share the same UID for the same role name.
+_ROLES_CACHE = {}
+
+
+def role(
+    adapter: SourceAdapter,
+    source_content_type: ContentTypeStr,
+    nautobot_name: FieldName = "role",
+) -> SourceFieldDefinition:
     """Create a role field definition.
 
     Use, when there is a different source role content type, that should be mapped to Nautobot "extras.role".
     Creates a new wrapper for the `source_content_type`, if it does not exist.
     """
+
+    def cache_roles(source: RecordData, importer_pass: ImporterPass) -> PreImportResult:
+        if importer_pass == ImporterPass.DEFINE_STRUCTURE:
+            name = source.get("name", "").capitalize()
+            if not name:
+                raise ValueError("Role name is required")
+            uid = _ROLES_CACHE.get(name, None)
+            nautobot_uid = role_wrapper.cache_record_uids(source, uid)
+            if not uid:
+                _ROLES_CACHE[name] = nautobot_uid
+
+        return PreImportResult.USE_RECORD
+
     role_wrapper = adapter.configure_model(
         source_content_type,
         nautobot_content_type="extras.role",
+        pre_import=cache_roles,
+        identifiers=("name",),
         fields={
             # Include color to allow setting the default Nautobot value, import fails without it.
             "color": "color",
         },
     )
 
-    return relation(role_wrapper, "role")
+    def define_role(field: SourceField) -> None:
+        def role_importer(source: RecordData, target: DiffSyncBaseModel) -> None:
+            value = field.get_source_value(source)
+            if value in EMPTY_VALUES:
+                return
+
+            if isinstance(value, int):
+                uid = role_wrapper.get_pk_from_uid(value)
+            elif isinstance(value, str):
+                value = value.capitalize()
+                uid = _ROLES_CACHE[value] if value in _ROLES_CACHE else role_wrapper.get_pk_from_identifiers([value])
+                role_wrapper.import_record({"id": uid, "name": value})
+            elif isinstance(value, UUID):
+                uid = value
```

Comment:
What are the data cases where we'd hit each of these branches? Would you mind adding some comments documenting this?

Reply:
Updated docstring, typing and added comments. Added caching, when referenced by a role name.